org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/hover/AbstractProblemHover.java

```diff
@@ -67,38 +68,81 @@ protected IAnnotationModel getAnnotationModel() {
+	 * line before, etc.
+	 */
 	public List<Annotation> getAnnotations(final int lineNumber, final int offset) {
 		if (getAnnotationModel() == null) {
 			return Collections.emptyList();
 		}
-		final Iterator<?> iterator = getAnnotationModel().getAnnotationIterator();
+		final Iterator<Annotation> iterator = getAnnotationModel().getAnnotationIterator();
 		List<Annotation> result = Lists.newArrayList();
+		List<Integer> lineDiffs = Lists.newArrayList();
 		while (iterator.hasNext()) {
-			final Annotation annotation = (Annotation) iterator.next();
-			if (isHandled(annotation)) {
+			final Annotation annotation = iterator.next();
+			if (isHandled(annotation) && !isLineDiffInfo(annotation)) {
 				Position position = getAnnotationModel().getPosition(annotation);
 				if (position != null) {
 					final int start = position.getOffset();
 					final int end = start + position.getLength();
 
-					if (offset > 0 && !(start <= offset && offset <= end)) {
-						continue;
-					}
-					try {
-						if (lineNumber != getDocument().getLineOfOffset(start)) {
-							continue;
+					if (offset < 0 || start <= offset && offset <= end) {
+						try {
+							int lineDiff = Math.abs(lineNumber - getDocument().getLineOfOffset(start));
+							if (lineDiff == 0 || offset >= 0) {
+								// Insert the annotation according to its line number difference
+								int index = lineDiffs.size();
+								while (index > 0 && lineDiffs.get(index - 1) > lineDiff) {
+									index--;
+								}
+								result.add(index, annotation);
+								lineDiffs.add(index, lineDiff);
+							}
+						} catch (Exception x) {
+							// Skip this annotation
 						}
-					} catch (final Exception x) {
-						continue;
-					}
-					if (!isLineDiffInfo(annotation)) {
-						result.add(annotation);
 					}
 				}
 			}
 		}
 		return result;
 	}
+	
+	/**
+	 * Sort the given annotations such that the resulting list contains first errors, then warnings, and then the rest.
+	 */
+	protected List<Annotation> sortBySeverity(List<Annotation> annotations) {
+		if (annotations == null)
+			return Collections.emptyList();
+		if (annotations.size() <= 1)
+			return annotations;
+		List<Annotation> sorted = Lists.newArrayList(annotations);
+		Collections.sort(sorted, new Comparator<Annotation>() {
+			@Override
+			public int compare(Annotation a1, Annotation a2) {
+				boolean isError1 = markerAnnotationAccess.isSubtype(a1.getType(), "org.eclipse.ui.workbench.texteditor.error");
+				boolean isError2 = markerAnnotationAccess.isSubtype(a2.getType(), "org.eclipse.ui.workbench.texteditor.error");
+				if (isError1 && isError2)
+					return 0;
+				if (isError1)
+					return -1;
+				if (isError2)
+					return 1;
+				boolean isWarning1 = markerAnnotationAccess.isSubtype(a1.getType(), "org.eclipse.ui.workbench.texteditor.warning");
+				boolean isWarning2 = markerAnnotationAccess.isSubtype(a2.getType(), "org.eclipse.ui.workbench.texteditor.warning");
+				if (isWarning1 && isWarning2)
```

Comment:
We should probably sort the annotations by message as the last argument to get a stable order.

Reply:
The sorting algorithm is stable, so it retains the order that was returned by `getAnnotations(int, int)` in case the annotations have the same type. That order is important since it puts those annotations first that start on the same line as the requested position in the document.

Previously we have just taken the order that was given by the annotation model. I haven't noticed any instability there.

Reply:
Good point, I wasn't sure about the default sorting of lists being stable.